src/firewall/firewalld.rs

```diff
@@ -60,19 +69,216 @@ impl firewall::FirewallDriver for FirewallD {
         Ok(())
     }
 
-    fn teardown_network(&self, _tear: TearDownNetwork) -> Result<(), Box<dyn Error>> {
-        todo!();
+    fn teardown_network(&self, tear: TearDownNetwork) -> Result<(), Box<dyn Error>> {
+        if !tear.complete_teardown {
+            return Ok(());
+        }
+
+        if let Some(subnets) = tear.config.net.subnets {
+            for subnet in subnets {
+                debug!("Removing subnet {} from zone {}", subnet.subnet, ZONENAME);
+                let _ = block_on(self.conn.call_method(
+                    Some("org.fedoraproject.FirewallD1"),
+                    "/org/fedoraproject/FirewallD1",
+                    Some("org.fedoraproject.FirewallD1.zone"),
+                    "removeSource",
+                    &(ZONENAME, subnet.subnet.to_string()),
+                ))?;
+            }
+        }
+
+        Ok(())
     }
 
-    fn setup_port_forward(&self, _setup_portfw: PortForwardConfig) -> Result<(), Box<dyn Error>> {
-        todo!();
+    fn setup_port_forward(&self, setup_portfw: PortForwardConfig) -> Result<(), Box<dyn Error>> {
+        // NOTE: There is a serious TOCTOU risk in this function if netavark
+        // is either run in parallel, or is not the only thing to edit this
+        // policy.
+        // Because of Podman's locking, this should be safe in the typical
+        // case.
+        // I don't think there's a safer way, unfortunately.
```

Comment:
Did you open a issue with firewalld to fix this API. I really do not like the TOCTOU risk here and it complicates the code versus just adding a new port.

Reply:
Also do you have to rewrite this to iptables now?

Reply:
Have not asked yet, but I will.

I'm going to leave this code intact but adjust the existing iptables code to allow iptables to also add appropriate rules to firewalld, if it exists, to allow traffic; then we can hardcode iptables-only for RHEL.